velox/functions/sparksql/tests/MapTest.cpp

```diff
@@ -55,55 +59,174 @@ TEST_F(MapTest, Nulls) {
+  testMapFails(
+      "map(c0, c1, c2)",
+      {inputVectorInt64, inputVectorDouble, inputVectorInt64},
+      "Scalar function signature is not supported: map(BIGINT, DOUBLE, BIGINT)");
+
+  testMapFails(
+      "map(c0, c1, c2, c3, c4, c5, c6, c7)",
+      {inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble},
+      "Scalar function signature is not supported: map(DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE, DOUBLE)");
 
   // Types of args
-  auto inputVector11 = makeNullableFlatVector<double>({10.0, 20.0, 30.0});
-  auto inputVector22 = makeNullableFlatVector<double>({4.1, 5.1, 6.1});
-  mapSimple(
+  testMapFails(
       "map(c0, c1, c2, c3)",
-      {inputVector1, inputVector2, inputVector11, inputVector22},
-      true);
+      {inputVectorInt64,
+       inputVectorDouble,
+       inputVectorDouble,
+       inputVectorDouble},
+      "Scalar function signature is not supported: map(BIGINT, DOUBLE, DOUBLE, DOUBLE)");
+  testMapFails(
+      "map(c0, c1, c2, c3)",
+      {inputVectorDouble,
+       inputVectorInt64,
+       inputVectorDouble,
+       inputVectorDouble},
+      "Scalar function signature is not supported: map(DOUBLE, BIGINT, DOUBLE, DOUBLE)");
+
+  testMapFails(
+      "map(c0, c1)",
+      {nullInputVector, inputVectorDouble},
+      "Cannot use null as map key");
+}
+
+TEST_F(MapTest, complexTypes) {
+  auto makeSingleMapVector = [&](const VectorPtr& keyVector,
+                                 const VectorPtr& valueVector) {
+    return makeMapVector(
+        {
+            0,
+        },
+        keyVector,
+        valueVector);
+  };
+
+  auto makeSingleRowVector = [&](vector_size_t size = 1,
+                                 vector_size_t base = 0) {
+    return makeRowVector({
+        makeFlatVector<int64_t>(size, [&](auto row) { return row + base; }),
+    });
+  };
+
+  auto arrayKey = makeArrayVectorFromJson<int64_t>({"[1, 2, 3]"});
+  auto arrayValue = makeArrayVectorFromJson<int64_t>({"[1, 3, 5]"});
+  auto nullArrayValue = makeNullableArrayVector<int64_t>({
+      std::nullopt,
+  });
+
+  testMap(
+      "map(c0, c1)",
+      {makeSingleRowVector(), makeSingleRowVector(1, 2)},
+      makeSingleMapVector(makeSingleRowVector(), makeSingleRowVector(1, 2)));
+
+  testMap(
+      "map(c0, c1)",
+      {arrayKey, arrayValue},
+      makeSingleMapVector(arrayKey, arrayValue));
+
+  testMap(
+      "map(c0, c1)",
+      {makeSingleMapVector(makeSingleRowVector(), makeSingleRowVector(1, 3)),
+       makeSingleMapVector(makeSingleRowVector(), makeSingleRowVector(1, 2))},
+      makeSingleMapVector(
+          makeSingleMapVector(makeSingleRowVector(), makeSingleRowVector(1, 3)),
+          makeSingleMapVector(
+              makeSingleRowVector(), makeSingleRowVector(1, 2))));
+
+  testMap(
```

Comment:
There is still a lot of code duplication. Maybe introduce testSingleMap(key, value)?

Reply:
Good suggestion, thanks.